hydromt/workflows/region.py

```diff
@@ -0,0 +1,287 @@
+
+        * {
+            'subbasin': [[x1, x2, ..], [y1, y2, ..]],
+            '<variable>': threshold, 'bounds': [xmin, ymin, xmax, ymax]
+            }
+
+        * {'subbasin': /path/to/point_geometry, '<variable>': threshold}
+
+        * {'subbasin': [xmin, ymin, xmax, ymax], '<variable>': threshold}
+
+        * {'subbasin': /path/to/polygon_geometry, '<variable>': threshold}
+
+        Interbasins are similar to subbasins but are bounded by a bounding box or
+        geometry and do not include all upstream area. Common use-cases include:
+
+        * {'interbasin': [xmin, ymin, xmax, ymax], '<variable>': threshold}
+
+        * {'interbasin': [xmin, ymin, xmax, ymax], 'xy': [x, y]}
+
+        * {'interbasin': /path/to/polygon_geometry, 'outlets': true}
+    """
+    kwargs = region.copy()
+    # NOTE: the order is important to prioritize the arguments
+    options = {
+        "basin": ["basid", "geom", "bbox", "xy"],
+        "subbasin": ["geom", "bbox", "xy"],
+        "interbasin": ["geom", "bbox", "xy"],
+        "geom": ["geom"],
+        "bbox": ["bbox"],
+        "mesh": ["UgridDataArray"],
+        "grid": ["raster"],
+    }
+    kind = next(iter(kwargs))  # first key of region
+    value0 = kwargs.pop(kind)
+
+    if kind in PLUGINS.model_plugins:
+        model_class = PLUGINS.model_plugins[kind]
+        other_model = model_class(root=value0, mode="r", logger=logger)
+        value0 = other_model.region
+        kwargs = dict(geom=other_model.region)
+        kind = "geom"
+    elif kind == "mesh":
+        if isinstance(value0, (str, Path)) and isfile(value0):
+            kwarg = dict(mesh=xu.open_dataset(value0))
+        elif isinstance(value0, (xu.UgridDataset, xu.UgridDataArray)):
+            kwarg = dict(mesh=value0)
+        elif isinstance(value0, (xu.Ugrid1d, xu.Ugrid2d)):
+            kwarg = dict(
+                mesh=xu.UgridDataset(value0.to_dataset(optional_attributes=True))
+            )
+        else:
+            raise ValueError(
+                f"Unrecognized type {type(value0)}."
+                "Should be a path, data catalog key or xugrid object."
+            )
+        kwargs.update(kwarg)
+    elif kind not in options:
+        k_lst = '", "'.join(list(options.keys()))
+        raise ValueError(f'Region key "{kind}" not understood, select from "{k_lst}"')
+
+    kwarg = _parse_region_value(value0, data_catalog=data_catalog)
+    if len(kwarg) == 0 or next(iter(kwarg)) not in options[kind]:
+        v_lst = '", "'.join(list(options[kind]))
+        raise ValueError(
+            f'Region value "{value0}" for kind={kind} not understood, '
+            f'provide one of "{v_lst}"'
+        )
+    kwargs.update(kwarg)
+
+    if kind in ["basin", "subbasin", "interbasin"]:
+        # retrieve global hydrography data (lazy!)
+        assert data_catalog is not None
+        assert hydrography_fn is not None
+        assert basin_index_fn is not None
```

Comment:
No mypy here.
Get_rasterdataset will take care of it if they are None.
AND basin_index_fn is actually allowed to be None!!

Reply:
I want to let mypy do its thing, otherwise there are red squiggles everywhere. I moved the assert of basin_index_fn inside the if statement.

And get_rasterdataset does not accept null as data_like